browser/net/brave_ad_block_tp_network_delegate_helper.cc

```diff
@@ -228,6 +235,50 @@ void UseCnameResult(scoped_refptr<base::SequencedTaskRunner> task_runner,
   }
 }
 
+// If only particular types of network traffic are being proxied, or if no
+// proxy is configured, it should be safe to continue making unproxied DNS
+// queries. However, in SingleProxy mode all types of network traffic should go
+// through the proxy, so additional DNS queries should be avoided.
+bool ProxySettingsAllowUncloaking(content::BrowserContext* browser_context) {
+  DCHECK(browser_context);
+
+  bool can_uncloak = true;
+
+  Profile* profile = Profile::FromBrowserContext(browser_context);
+
+  std::unique_ptr<PrefProxyConfigTracker> config_tracker =
+      ProxyServiceFactory::CreatePrefProxyConfigTrackerOfProfile(
+          profile->GetPrefs(), nullptr);
+  std::unique_ptr<net::ProxyConfigService> proxy_config_service =
+      ProxyServiceFactory::CreateProxyConfigService(config_tracker.get());
+
+  net::ProxyConfigWithAnnotation config;
+  net::ProxyConfigService::ConfigAvailability availability =
+      proxy_config_service->GetLatestProxyConfig(&config);
+
+  if (availability ==
+      net::ProxyConfigService::ConfigAvailability::CONFIG_VALID) {
+    LOG(ERROR) << "CONFIG_VALID";
+    // PROXY_LIST corresponds to SingleProxy mode.
+    if (config.value().proxy_rules().type ==
+        net::ProxyConfig::ProxyRules::Type::PROXY_LIST) {
+      LOG(ERROR) << "PROXY_LIST";
+      can_uncloak = false;
+    }
+  } else if (availability ==
+             net::ProxyConfigService::ConfigAvailability::CONFIG_PENDING) {
+    // Fallback to not CNAME uncloaking if the proxy configuration cannot be
+    // determined.
+    LOG(ERROR) << "CONFIG_PENDING";
+    can_uncloak = false;
+  }
```

Comment:
According to CI, this gets triggered on Windows without any additional configuration. It's unclear whether or not that's the case every time the proxy service is created, in which case we'll never have CNAME uncloaking on Windows with this approach, or if it will permanently be resolved after a short time on browser startup, in which case this is probably an acceptable fix.

Any thoughts @iefremov @darkdh?